Corrected is_tangent in ellipse.py

Amit-Modi opened the pull request:
What's Wrong: #19471
```python
>>> from sympy import *
>>> c = Circle(Point(0,0),2)
>>> poly1 = Polygon((-2,-2),(-2,2),(2,2),(3,0),(2,-2))
>>> c.is_tangent(poly1)
True
>>> poly2 = Polygon((2,2),(3,0),(2,-2),(-2,-2),(-2,2))
>>> c.is_tangent(poly2)
False
>>> r = Ray(Point(0,2),Point(2,2))
>>> c.is_tangent(r)
False
```

notice that poly1 and poly2 are giving different results,
but are same polygon except there starting points are different.
Further, in my opinion( I found very less on this topic on internet but by definition in google),  tangent is a straight line or plane that touches a curve or curved surface at a point, but if EXTENDED does not cross it at that point.
the ray,r touches the circle,c at r.source but when extended does not cross circle,c at any point. by this definition r is a tangent to circle c.

Changes I made:
change 1-
Line Number:720-730
       if one or more sides of polygon are tangent to ellipse and other does not touch it, polygon is a tangent
      

change 2-
Line Number: 714
       original content : 
                return intersect[0] != o.source and not self.encloses_point(o.source)
       my change:
                return o in self.tangent_lines(intersect[0])[0]

Amit-Modi committed:
Corrected is_tangent in ellipse.py

>>> from sympy import *
>>> c=Circle(Point(0,0),2)
>>> poly1=Polygon((-2,-2),(-2,2),(2,2),(3,0),(2,-2))
>>> c.is_tangent(poly1)
True
>>> poly2=Polygon((2,2),(3,0),(2,-2),(-2,-2),(-2,2))
>>> c.is_tangent(poly2)
False
>>> r=Ray(Point(0,2),Point(2,2))
>>> c.is_tangent(s)
False

notice that poly1 and poly2 are giving different results,
but are same polygon except there starting points are different.
Further, in my opinion( I found very less on this topic on internet but by definition in google),  tangent is a straight line or plane that touches a curve or curved surface at a point, but if EXTENDED does not cross it at that point.
the ray,r touches the circle,c at r.source but when extended does not cross circle,c at any point. by this definition r is a tangent to circle c.

certik commented:
Hi @Amit-Modi, thanks for the fix. Can you please also write a test, that failed before, but now it works? That would be the case for `r` that you posted above.

Amit-Modi commented:
Hi @certik, thanks for your help. I tried to do write a test, but it failed. Can you help me to point out my mistake.

certik commented:
I think the error is unrelated. I restarted the failing test.

gxyd commented:
Can you add a mention a few reference for the definition of this? For like definition of intersection of Ray with Circle/Ellipse. Also for `Segment`.

gxyd commented:
Can you add a mention a few reference for the definition of this? For like definition of intersection of Ray with Circle/Ellipse. Also for `Segment`.
Add them in the file as well.

[Edit]: I mistakenly deleted the last comment.

Amit-Modi commented:
@gxyd, sir the following definitions are taken from [http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html](http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html)
Definition: A tangent to a circle is a line in the plane of the circle that intersects the circle in exactly one point. The point of intersection is called the point of tangency.
![unitstdy4](https://user-images.githubusercontent.com/25955291/32133131-610a8a02-bbef-11e7-9da2-3d824470cc3e.gif)
In the figure, line DC is a tangent; point B is the point of tangency. A ray or segment is tangent if it is a part of a tangent line and contains the point of tangency. So, ray BC is a tangent ray and segment BC is a tangent segment.

In which file I add this ?

Amit-Modi committed:
add tests to is_tangent  test_ellipse.py

added tangent ray tests and tangent polygon tests to is_tangent  test_ellipse.py

ylemkimon commented:
Please add it to `References` section like below in the docstring. See docstrings of other functions or classes for reference.
```
    References
    ==========
```

Amit-Modi committed:
added Reference to docstring of is_tangent

a reference to A UNIT OF STUDY in GEOMETRY by TERESA BANKER added to docstring for the definition of tangent ray and tangent segment.<br>
Definition for tangent ray and tangent segment are mentioned under lesson 2.

Amit-Modi commented:
Thanks @ylemkimon, for your help. <br>And please tell me if anything else is required.

jksuom commented on sympy/geometry/ellipse.py:
```diff
@@ -694,6 +694,11 @@ def is_tangent(self, o):
+        References
+        ==========
+
+        [1] http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html
```
This `student.folders` link is likely to disappear in a short while. Would it be possible to find a more stable reference?

Amit-Modi replied:
@ylemkimon, I tried to find a more stable reference.<br> I mentioned earlier, I did not find a definition for tangent ray/segment (other than this ). <br>In most of the places tangent is defined as a tangent line, but I found following definition at (https://en.wikipedia.org/wiki/Tangent_(disambiguation))[https://en.wikipedia.org/wiki/Tangent_(disambiguation)]<br><br>More generally, in geometry, two curves are said to be tangent when they intersect at a given point and have the same direction at that point

jksuom commented:
It seems to me that Change 2 is correct but Change 1 may be too strict. Two curves have generally several points in common, and they may be tangent to each other at some, but at not all of them. Hence it is usually said that the curves are tangent *at a certain point* instead of just saying that they are tangent. It is also sometimes said that the curves are tangent to each other if they are tangent at one point or several points.

It is also possible that two curves cross each other at a point where they have a common tangent. For example, consider the curve `y = x**3` in the `x-y`-plane that is tangent to the `x`-axis at the origin. 

It appears hard to find an easily accessible reference for the concept of tangency. Perhaps none would be necessary.

gxyd commented:
@Amit-Modi can you please address the requested changes?

vishal-vardhan commented:
@certik @gxyd @ylemkimon @jksuom can i work on this issue?

Amit-Modi commented:
moved to #15774

czgdp1807 commented:
@Amit-Modi Is this issue resolved? I don't think no PR other than #15774 (which is open) referred this issue. Can you please reopen this issue. Thanks.

smichr commented on sympy/geometry/tests/test_ellipse.py:
```diff
@@ -378,6 +379,7 @@ def test_is_tangent():
     assert e1.is_tangent(Polygon((-3, 0), (3, 0), (0, 5))) is False
     assert e1.is_tangent(Polygon((-3, 0), (0, -5), (3, 0), (0, 5))) is False
     assert e1.is_tangent(Polygon((-3, -5), (-3, 5), (3, 5), (3, -5))) is True
+    assert e1.is_tangent(Polygon((3, 5), (3, -5), (-3, -5), (-3, 5))) is True
```
this has been fixed in #24716

smichr commented on sympy/geometry/tests/test_ellipse.py:
```diff
@@ -362,6 +362,7 @@ def test_is_tangent():
     assert c1.is_tangent(Ray((-3, -2), (-15, -20))) is False
     assert c1.is_tangent(Ray((-3, -22), (15, 20))) is False
     assert c1.is_tangent(Ray((9, 20), (9, -20))) is True
+    assert c1.is_tangent(Ray((2, 5), (9, 5))) is True
```
this behavior is still the same at this time.
